File: filebeat/scripts/generate_fields_yml.py
import json
import re
import yaml

from argparse import ArgumentParser
from collections import OrderedDict


FIELDS_YML_PATH = "module/{module}/{fileset}/_meta/fields.yml"
PIPELINE_JSON_PATH = "module/{module}/{fileset}/ingest/pipeline.json"
DEFAULT_TYPE = "text"
TYPES = {
    "WORD": "keyword",
    "USERNAME": "keyword",
    "HOSTNAME": "keyword",
    "LOGLEVEL": "keyword",
    "GREEDYDATA": "text",
    "GREEDYMULTILINE": "text",
    "IPHOST": "keyword",
    "IPORHOST": "keyword",
    "NUMBER": "long",
}
FUNCTIONS = {
    "remove": None,
    "rename": None,
}


def _get_elements_of_grok_patterns(pipeline):
    elements = []
    patterns = pipeline["processors"][0]["grok"]["patterns"]

    for pattern in patterns:
        fields = re.findall("{[\.\w\:]*}", pattern)
        for field in fields:
            field_type, name = field[1:-1].split(":")
            elements.append((field_type, name))

    return elements


def get_message_elements_from_pipeline(module_name, fileset_name):
    pipeline_path = PIPELINE_JSON_PATH.format(module=module_name, fileset=fileset_name)
    with open(pipeline_path, "r") as pipeline_file:
        pipeline = json.load(pipeline_file, object_pairs_hook=OrderedDict)

    return _get_elements_of_grok_patterns(pipeline)


def __get_field_by_name(fields, field_name):
    for field in fields:
        if "name" in field and field["name"] == field_name:
            return field
    return None


def __insert_last_field(fields, element_type, element_name):
    field = __get_field_by_name(fields, element_name)
    if field:
        return

    field_type = DEFAULT_TYPE
    if element_type in TYPES:
        field_type = TYPES[element_type]
    new_field = {
                    "name": element_name,
                    "type": field_type,
                    "description": "Please add description",
                    "example": "Please provide an example",
                }
    fields.append(new_field)


def __insert_into_existing_group(field, field_elements, element_type, index, count):
    _insert_fields_into_dict(field["fields"], field_elements, element_type, index+1, count)


def __insert_new_group(fields, field_elements, element_type, index, count):
    its_fields = []
    _insert_fields_into_dict(its_fields, field_elements, element_type, index+1, count)

    if its_fields:
        fields.append({
            "name": field_elements[index],
            "type": "group",
            "description": "Please add description",
            "fields": its_fields,
        })


def __insert_group(fields, field_elements, element_type, index, count):
    field = __get_field_by_name(fields, field_elements[index])
    if field is not None:
        __insert_into_existing_group(field, field_elements, element_type, index, count)
    else:
        __insert_new_group(fields, field_elements, element_type, index, count)


def _insert_fields_into_dict(fields, field_elements, element_type, index, count):
    if index+1 == count:
        __insert_last_field(fields, element_type, field_elements[index])
        return
    __insert_group(fields, field_elements, element_type, index, count)


def get_nested_fields(message_elements):
    fields = []
    for element_type, name in message_elements:
        name_elements = name.split(".")
        elements_count = len(name_elements)
        _insert_fields_into_dict(fields, name_elements, element_type, 1, elements_count)

    return fields


def save_yml(fields_yml, module_name, fileset_name):
    fields_yml_path = FIELDS_YML_PATH.format(module=module_name, fileset=fileset_name)
    with open(fields_yml_path, "w") as out:
        yaml.safe_dump(fields_yml, out, default_flow_style=False, encoding="utf-8", allow_unicode=True, width=100)


def generate_fields_yml(module_name, fileset_name):
    message_elements = get_message_elements_from_pipeline(module_name, fileset_name)
    fields_yml = get_nested_fields(message_elements)
    save_yml(fields_yml, module_name, fileset_name)


if __name__ == "__main__":
    parser = ArgumentParser(description="Generate fields.yml for new Filebeat modules from pipeline.json")
    parser.add_argument("module")
    parser.add_argument("fileset")
    args = parser.parse_args()

    generate_fields_yml(args.module, args.fileset)

File: filebeat/fileset/generate_fields_yml.go
package fileset

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"reflect"
	"regexp"
	"strings"

	"gopkg.in/yaml.v2"
)

const (
	pipelinePath  = "../module/%s/%s/ingest/pipeline.json"
	fieldsYmlPath = "../module/%s/%s/_meta/fields.yml"
)

var (
	types = map[string]string{
		"group":           "group",
		"DATA":            "text",
		"GREEDYDATA":      "text",
		"GREEDYMULTILINE": "text",
		"HOSTNAME":        "keyword",
		"IPHOST":          "keyword",
		"IPORHOST":        "keyword",
		"LOGLEVEL":        "keyword",
		"MULTILINEQUERY":  "text",
		"NUMBER":          "long",
		"POSINT":          "long",
		"SYSLOGHOST":      "keyword",
		"SYSLOGTIMESTAMP": "text",
		"TIMESTAMP":       "text",
		"USERNAME":        "keyword",
		"WORD":            "keyword",
	}
	// Nodoc specifies if the generated field.yml is includes documentation fields e.g. description, example
	Nodoc bool
)

type pipeline struct {
	Description string                   `json:"description"`
	Processors  []map[string]interface{} `json:"processors"`
	OnFailure   interface{}              `json:"on_failure"`
}

type field struct {
	Type     string
	Elements []string
}

type fieldYml struct {
	Name        string      `yaml:"name"`
	Description string      `yaml:"description,omitempty"`
	Example     string      `yaml:"example,omitempty"`
	Type        string      `yaml:"type,omitempty"`
	Fields      []*fieldYml `yaml:"fields,omitempty"`
}

func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
	if noDoc {
		return &fieldYml{
			Name: name,
			Type: typeName,
		}
	}

	return &fieldYml{
		Name:        name,
		Type:        typeName,
		Description: "Please add description",
		Example:     "Please add example",
	}
}

func newField(lp string) field {
	lp = lp[1 : len(lp)-1]
	ee := strings.Split(lp, ":")
	e := strings.Split(ee[1], ".")
	return field{
		Type:     ee[0],
		Elements: e,
	}
}

func readPipeline(module, fileset string) (*pipeline, error) {
	pp := fmt.Sprintf(pipelinePath, module, fileset)
	r, err := ioutil.ReadFile(pp)
	if err != nil {
		return nil, err
	}

	var p pipeline
	err = json.Unmarshal(r, &p)
	if err != nil {
		return nil, err
	}

	return &p, nil
}

func addNewField(fs []field, f field) []field {
	for _, ff := range fs {
		if reflect.DeepEqual(ff, f) {
			return fs
		}
	}
	return append(fs, f)
}

func getElementsFromPatterns(patterns []string) ([]field, error) {
	r, err := regexp.Compile("{[\\.\\w\\:]*}")
	if err != nil {
		return nil, err
	}

	fs := make([]field, 0)
	for _, lp := range patterns {
		pp := r.FindAllString(lp, -1)
		for _, p := range pp {
			f := newField(p)
			fs = addNewField(fs, f)
		}

	}
	return fs, nil
}

func accumulatePatterns(grok interface{}) ([]string, error) {
	for k, v := range grok.(map[string]interface{}) {
		if k == "patterns" {
			vs := v.([]interface{})
			p := make([]string, 0)
			for _, s := range vs {
				p = append(p, s.(string))
			}
			return p, nil
		}
	}
	return nil, fmt.Errorf("No patterns in pipeline")
}

func accumulateRemoveFields(remove interface{}, out []string) []string {
	for k, v := range remove.(map[string]interface{}) {
		if k == "field" {
			vs := v.(string)
			return append(out, vs)
		}
	}
	return out
}

func accumulateRenameFields(rename interface{}, out map[string]string) map[string]string {
	var from, to string
	for k, v := range rename.(map[string]interface{}) {
		if k == "field" {
			from = v.(string)
		}
		if k == "target_field" {
			to = v.(string)
		}
	}
	out[from] = to
	return out
}

type processors struct {
	patterns []string
	remove   []string
	rename   map[string]string
}

func (p *processors) processFields() ([]field, error) {
	f, err := getElementsFromPatterns(p.patterns)
	if err != nil {
		return nil, err
	}

	for i, ff := range f {
		fs := strings.Join(ff.Elements, ".")
		for _, rm := range p.remove {
			if fs == rm {
				f = append(f[:i], f[i+1:]...)
			}
		}
		for k, mv := range p.rename {
			if k == fs {
				ff.Elements = strings.Split(mv, ".")
			}
		}
		f[i] = ff
	}
	return f, nil
}

func getProcessors(p []map[string]interface{}) (*processors, error) {
	patterns := make([]string, 0)
	rmFields := make([]string, 0)
	mvFields := make(map[string]string)
	var err error

	for _, e := range p {
		if ee, ok := e["grok"]; ok {
			patterns, err = accumulatePatterns(ee)
			if err != nil {
				return nil, err
			}
		}
		if rm, ok := e["remove"]; ok {
			rmFields = accumulateRemoveFields(rm, rmFields)
		}
		if mv, ok := e["rename"]; ok {
			mvFields = accumulateRenameFields(mv, mvFields)
		}
	}

	if patterns == nil {
		return nil, fmt.Errorf("No patterns in pipeline")
	}

	return &processors{
		patterns: patterns,
		remove:   rmFields,
		rename:   mvFields,
	}, nil
}

func getFieldByName(f []*fieldYml, name string) *fieldYml {
	for _, ff := range f {
		if ff.Name == name {
			return ff
		}
	}
	return nil
}

func insertLastField(f []*fieldYml, name, typeName string, noDoc bool) []*fieldYml {
	ff := getFieldByName(f, name)
	if ff != nil {
		return f
	}

	nf := newFieldYml(name, types[typeName], noDoc)
	return append(f, nf)
}

func insertGroup(out []*fieldYml, field field, index, count int, noDoc bool) []*fieldYml {
	g := getFieldByName(out, field.Elements[index])
	if g != nil {
		g.Fields = generateField(g.Fields, field, index+1, count, noDoc)
		return out
	}

	groupFields := make([]*fieldYml, 0)
	groupFields = generateField(groupFields, field, index+1, count, noDoc)
	group := newFieldYml(field.Elements[index], "group", noDoc)
	group.Fields = groupFields
	return append(out, group)
}

func generateField(out []*fieldYml, field field, index, count int, noDoc bool) []*fieldYml {
	if index+1 == count {
		return insertLastField(out, field.Elements[index], field.Type, noDoc)
	}
	return insertGroup(out, field, index, count, noDoc)
}

func generateFields(f []field, noDoc bool) []*fieldYml {
	out := make([]*fieldYml, 0)
	for _, ff := range f {
		out = generateField(out, ff, 1, len(ff.Elements), noDoc)
	}
	return out
}

func (p *pipeline) toFieldsYml(noDoc bool) ([]byte, error) {
	pr, err := getProcessors(p.Processors)
	if err != nil {
		return nil, err
	}

	var fs []field
	fs, err = pr.processFields()
	if err != nil {
		return nil, err
	}

	f := generateFields(fs, noDoc)
	var d []byte
	d, err = yaml.Marshal(&f)

	return d, nil
}

func writeFieldsYml(module, fileset string, f []byte) error {
	p := fmt.Sprintf(fieldsYmlPath, module, fileset)
	err := ioutil.WriteFile(p, f, 0664)
	if err != nil {
		return err
	}
	return nil
}

// GenerateFieldsYml creates a fields.yml file for fileset based on the existing pipeline.json
func GenerateFieldsYml(module, fileset string) error {
	p, err := readPipeline(module, fileset)
	if err != nil {
		return err
	}

	var d []byte
	d, err = p.toFieldsYml(Nodoc)
	if err != nil {
		return err
	}

	err = writeFieldsYml(module, fileset, d)
	if err != nil {
		return err
	}

	return nil
}
